feat(avm): Use More Multipermutations

MirandaWood opened the pull request:
### More Multiperms

This PR integrates multipermutations for the public data tree check trace and bytecode hashing trace. It also covers more bytecode tasks listed here: #17428 (the removal of the hash correctness lookup is hopefully justified in a comment on that issue!).

Notes/key points:

- The public data tree perms are split into `public_data_check_write -> public_data_check_protocol_write + public_data_check_non_protocol_write`
- The bytecode field perms are split into `bc_decomposition_sel_packed -> bc_decomposition_sel_packed_read_0 _ + bc_decomposition_sel_packed_read_1 _ + bc_decomposition_sel_packed_read_2 _`
- Since `InteractionDefinition.add()` names multipermutations by the concatenation of each perm's name, `check_interaction` will not work, so I had to remove a couple of tests. I can add a hacky method to get these working again if we like!

MirandaWood commented on barretenberg/cpp/pil/vm2/bytecode/bc_decomposition.pil:
```diff
@@ -230,13 +230,27 @@ bytes_pc_plus_36 = (1 - FIRST_OR_LAST_CONTRACT) * bytes_pc_plus_35';
 
 // TODO: We need to ensure that every active row contains a byte which is accounted for
 // in the bytecode hashing trace. See https://github.com/AztecProtocol/aztec-packages/issues/17428
+// TODO(MW): Is the above now covered by the perm between hashing and decomp?
```
It was hard to get my head around whether the permutation between hashing and decomp plus the use of shifts for each byte in this trace now means we can remove the above `todo` - any thoughts? I believe we can?

MirandaWood replied:
Now that we constrain in decomp that `sel_packed = 1` every 31st byte, and each shifted byte (i.e. `bytes_pc_plus_i` for `i` up to 30) is included in `packed_field` which is in the multiperm, I think this todo has been resolved?

jeanmon replied:
Sure, let us remove this TODO.

jeanmon commented on barretenberg/cpp/pil/vm2/bytecode/bc_decomposition.pil:
```diff
@@ -230,13 +230,27 @@ bytes_pc_plus_36 = (1 - FIRST_OR_LAST_CONTRACT) * bytes_pc_plus_35';
 
+// Each first row of a new bytecode id should be packed (#17428):
+#[SEL_PACKED_INIT]
+FIRST_OR_LAST_CONTRACT * (sel' - sel_packed') = 0;
```
It is a bit outside of the scope of this PR but would you mind renaming FIRST_OR_LAST_CONTRACT to "LATCH_CONDITION" which is a more standard naming through our code base?

jeanmon commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -242,6 +242,10 @@ namespace bc_hashing;
     #[HASH_CONSISTENCY]
     (1 - LATCH_CONDITION) * (output_hash' - output_hash) = 0;
 
+    // Since we constrain both hash and id consistency, it is sufficient to check only on the final row:
```
Would you mind renaming #[ID_CONSISTENCY] and  #[HASH_CONSISTENCY] by  #[ID_PROPAGATION] and  #[HASH_PROPAGATION] and adapting the comment accordingly?

I think we actually do not need #[HASH_CONSISTENCY]. WDYT?

MirandaWood replied:
I think if we remove `#[HASH_CONSISTENCY]` it's possible to set an incorrect hash at some non-latch rows and match them up to 'unrelated' poseidon rows. We could check `#[HASH_IS_ID]` on every activated row however and remove `#[HASH_CONSISTENCY]`?

jeanmon replied:
My bad. Yes I think you are right that Poseidon needs this. Maybe we can add a comment on this.

jeanmon replied:
Great Idea @MirandaWood  GO ahead with removing #[HASH_CONSISTENCY] and activate #[HASH_IS_ID] at every row. Please add a comment about why it is crucial ti ensure this at every row (due to Poseidon).

jeanmon commented on barretenberg/cpp/pil/vm2/opcodes/sstore.pil:
```diff
@@ -1,4 +1,5 @@
 include "../constants_gen.pil";
+include "../precomputed.pil";
```
Was this import added inadvertently?

MirandaWood replied:
No - we do use precomputed in this file, but it was not explicitly imported (it ran because an import imports it, I believe). I've previously been asked to explicity import so thought I would do so here!

jeanmon replied:
I did not know that import of import were working in .pil. Good to keep this import then.

MirandaWood commented on barretenberg/cpp/src/barretenberg/vm2/constraining/testing/check_relation.hpp:
```diff
@@ -78,6 +78,15 @@ template <typename TraceBuilder, typename... Setting> inline void check_interact
     (TraceBuilder::interactions.template get_test_job<Setting>()->process(trace), ...);
 }
 
+// Warning: The below requires ALL permutation settings as defined in InteractionDefinition.add():
+template <typename TraceBuilder, typename... Setting>
+inline void check_multipermutation_interaction(tracegen::TestTraceContainer& trace)
+{
+    // Concatenates the names of given permutation interactions:
+    std::string name = (std::string(Setting::NAME) + ...);
+    TraceBuilder::interactions.get_test_job(name)->process(trace);
+}
+
```
This is a bit hacky, so I'm happy to remove the revert the commit which added this one!

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/bc_hashing.test.cpp:
```diff
@@ -213,10 +213,13 @@ TEST_F(BytecodeHashingConstrainingTestTraceHelper, SingleBytecodeHashMax)
     GTEST_SKIP();
```
Maybe we can un-skip it by now.

MirandaWood replied:
It still takes around 4s, but I will check again after the next changes!

MirandaWood replied:
It now takes around 5s - I think the increase is expected now that we must do some more work to check and set `next_packed_pc` in the decomp trace, so I'll keep it skipped for now?

jeanmon replied:
At this stage of development, it is fine to re-enable. We have a couple of tests taking more than 10 seconds.

MirandaWood replied:
With batch inverses it's down to 2s, so will definitely re-add, thanks!

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/bc_hashing.test.cpp:
```diff
@@ -256,17 +262,29 @@ TEST_F(BytecodeHashingConstrainingTest, BytecodeInteractions)
-        { { .bytecode_id = 1, .bytecode = std::make_shared<std::vector<uint8_t>>(bytecode) } }, trace);
+        { { .bytecode_id = hash, .bytecode = std::make_shared<std::vector<uint8_t>>(bytecode) } }, trace);
+
+    // TODO(MW): Not currently used, but useful for if/when ultiperm interactions are testable via check_interaction
```
```suggestion
    // TODO(MW): Not currently used, but useful for if/when multiperm interactions are testable via check_interaction
```

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/lib/interaction_def.cpp:
```diff
@@ -33,6 +33,8 @@ std::unique_ptr<InteractionBuilderInterface> InteractionDefinition::get_test_job
     return get_job_internal(interaction_name)(/*strict=*/true);
 }
 
+// TODO(MW): get_multiperm_test_job? Multipermutations are listed by concatentated names (see nteraction_def.hpp)
```
```suggestion
// TODO(MW): get_multiperm_test_job? Multipermutations are listed by concatentated names (see interaction_def.hpp)
```

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/public_data_tree_trace.cpp:
```diff
@@ -249,6 +256,12 @@ const InteractionDefinition PublicDataTreeTraceBuilder::interactions =
         .add<lookup_public_data_check_new_leaf_merkle_check_settings, InteractionType::LookupGeneric>()
+        .add<InteractionType::MultiPermutation, perm_sstore_storage_write_settings, perm_tx_balance_update_settings>(
+            Column::public_data_check_write)
+        // .add<perm_sstore_storage_write_settings, InteractionType::MultiPermutation>(
```
You can probably delete these comments.

jeanmon reviewed (changes requested):
Good work. Can you clarify with @fcarreiro if there will be a possibility to test multi-permutation in the future in a unit test? If no, I would remove all the related comments.

jeanmon commented on barretenberg/cpp/pil/vm2/bytecode/bc_decomposition.pil:
```diff
@@ -167,76 +167,102 @@ last_of_contract * (1 - sel_windows_gt_remaining) = 0;
 // For bytecode hashing, we need to re-pack 31 bytes at some PCs into a field.
 // We will have a selector for the PCs that are packed. This only needs to happen
 // for PCs 0, 31, 62, ...
 // The "sequencer" can choose which PCs to pack, but the bytecode hashing trace
```
This text here should be removes or updated as the sequencer cannot choose which PC to pack anymore.

MirandaWood replied:
Ah missed this, thanks!

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/bc_decomposition.test.cpp:
```diff
@@ -34,6 +36,31 @@ void init_trace(TestTraceContainer& trace)
+    // perm_builder.process(trace);
+    // ...But since we are only testing decomp (without hashing) in some cases, this helper sets the selectors so we
+    // don't fail:
+    std::vector<C> selectors = { C::bc_decomposition_sel_packed_read_0_,
```
nitpick: could be a constexpr array

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/bc_decomposition.test.cpp:
```diff
@@ -34,6 +36,31 @@ void init_trace(TestTraceContainer& trace)
+                                 C::bc_decomposition_sel_packed_read_1_,
+                                 C::bc_decomposition_sel_packed_read_2_ };
+    uint32_t j = 0;
+    for (uint32_t r = 0; r < trace.get_num_rows(); r++) {
```
Difficult to write my head around this code. What could help IMHO:
- remove r++ in the loop and just use `r += bytes_remaining > 0 ? bytes_remaining : 1;` at the end of the loop body.
- remove variable `j` and inline `selectors[((i -r) % 31) % 3]`

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/bc_decomposition.test.cpp:
```diff
@@ -465,5 +496,94 @@ TEST(BytecodeDecompositionConstrainingTest, NegativeSelPackedNotSel)
+    set_perm_selectors(trace);
+    check_relation<bc_decomposition>(trace);
+
+    // At row 32, pc = 31 and should be packed:
```
Could you add an ASSERT_EQ(pc, 31) ?

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/bytecode_trace.cpp:
```diff
@@ -124,7 +126,16 @@ void BytecodeTraceBuilder::process_decomposition(
+                trace.set(row + j,
+                          { {
+                              { C::bc_decomposition_next_packed_pc, i + 31 },
+                              { C::bc_decomposition_next_packed_pc_min_pc_inv, (FF(i + 31) - FF(j)).invert() },
```
@fcarreiro just merged batch field inversion in a PR which was just merged. Could you add this batching for this one as well?

jeanmon reviewed (approved):
High-quality work!

MirandaWood committed:
feat(avm): Use More Multipermutations

### More Multiperms

This PR integrates multipermutations for the public data tree check trace and bytecode hashing trace. It also covers more bytecode tasks listed here: #17428 (the removal of the hash correctness lookup is hopefully justified in a comment on that issue!).

Notes/key points:

- The public data tree perms are split into `public_data_check_write -> public_data_check_protocol_write + public_data_check_non_protocol_write`
- The bytecode field perms are split into `bc_decomposition_sel_packed -> bc_decomposition_sel_packed_read_0 _ + bc_decomposition_sel_packed_read_1 _ + bc_decomposition_sel_packed_read_2 _`
- Since `InteractionDefinition.add()` names multipermutations by the concatenation of each perm's name, `check_interaction` will not work, so I had to remove a couple of tests. I can add a hacky method to get these working again if we like!
